src/server/snap.rs

```diff
@@ -239,16 +239,38 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
                 SNAP_TASK_COUNTER.with_label_values(&["write"]).inc();
                 match self.files.entry(token) {
                     Entry::Occupied(mut e) => {
-                        if let Err(err) = data.write_all_to(&mut e.get_mut().0) {
-                            error!(
-                                "failed to write data to snapshot file {} for token {:?}: {:?}",
-                                e.get_mut().0.path(),
-                                token,
-                                err
-                            );
-                            let (_, msg) = e.remove();
-                            let key = SnapKey::from_snap(msg.get_message().get_snapshot()).unwrap();
-                            self.snap_mgr.deregister(&key, &SnapEntry::Receiving);
+                        let mut finish = true;
+                        {
+                            let (left, right) = data.slice();
+                            let limiter = self.snap_mgr.get_limiter();
+                            if let Err(err) = limiter.write(&mut e.get_mut().0, left) {
```

Comment:
if you write left, you should also write right at the same time. 

I suggest extracting the function to avoid duplicated error handling.

Reply:
But I think the better way is to implement a LimitWriter, so you don't need to change these codes here.

```
let mut w = LimitWriter{limiter, &e.get_mut().0};
if let Err(err) = data.write_all_to(&w)
```

Reply:
I finally understand this.